contracts/contracts/coordination/Subscription.sol

```diff
@@ -0,0 +1,261 @@
+import "./Coordinator.sol";
+
+using SafeERC20 for IERC20;
+
+/**
+ * @title Subscription
+ * @notice Manages the subscription information for rituals.
+ * @dev This contract is abstract and should be extended by a concrete implementation.
+ * It maintains a reference to a Coordinator contract and a fee token (ERC20).
+ * Each subscription has an associated SubscriptionInfo struct which keeps track of the subscription details.
+ */
+abstract contract Subscription {
+    struct SubscriptionInfo {
+        uint256 paidFor;
+        uint256 spent;
+        uint256 expiration;
+        address subscriber;
+    }
+
+    Coordinator public coordinator;
+    IERC20 public feeToken;
+
+    // Mapping from subscription ID to subscription info
+    mapping(uint32 => SubscriptionInfo) public subscriptions;
+
+    // Mapping from (ritualId, address) to subscription ID
+    mapping(bytes32 => uint32) public subscribers;
+
+    uint32 public numberOfSubscriptions;
+
+    // TODO: DAO Treasury
+    address public beneficiary;
+
+    /**
+     * @notice Emitted when a subscription is created
+     * @param subscriptionId The ID of the subscription
+     * @param subscriber The address of the subscriber
+     * @param ritualId The ID of the ritual
+     */
+    event SubscriptionCreated(
+        uint32 indexed subscriptionId,
+        address indexed subscriber,
+        uint32 indexed ritualId
+    );
+
+    /**
+     * @notice Emitted when a subscription is cancelled
+     * @param subscriptionId The ID of the subscription
+     * @param subscriber The address of the subscriber
+     * @param ritualId The ID of the ritual
+     */
+    event SubscriptionCancelled(
+        uint32 indexed subscriptionId,
+        address indexed subscriber,
+        uint32 indexed ritualId
+    );
+
+    /**
+     * @notice Emitted when a subscription is paid
+     * @param subscriptionId The ID of the subscription
+     * @param subscriber The address of the subscriber
+     * @param amount The amount paid
+     */
+    event SubscriptionPaid(
+        uint32 indexed subscriptionId,
+        address indexed subscriber,
+        uint256 amount
+    );
+
+    /**
+     * @notice Sets the coordinator and fee token contracts
+     * @dev The coordinator and fee token contracts cannot be zero addresses
+     * @param _coordinator The address of the coordinator contract
+     * @param _feeToken The address of the fee token contract
+     * @param _beneficiary The address of the beneficiary
+     */
+    constructor(Coordinator _coordinator, IERC20 _feeToken, address _beneficiary) {
+        require(address(_coordinator) != address(0), "Coordinator cannot be the zero address");
+        require(address(_feeToken) != address(0), "Fee token cannot be the zero address");
+        require(_beneficiary != address(0), "Beneficiary cannot be the zero address");
+        coordinator = _coordinator;
+        feeToken = _feeToken;
+        beneficiary = _beneficiary;
```

Comment:
can we make these addresses immutable? at least coordinator and feeToken, unless they can be changed without upgrade in the future